keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationSupport.kt

```diff
@@ -18,59 +18,178 @@
+ *     should check the caller has access to the service account, in addition to WRITE access to the application.
+ */
 @Component
 class AuthorizationSupport(
   private val permissionEvaluator: FiatPermissionEvaluator,
-  private val repository: KeelRepository
+  private val repository: KeelRepository,
+  private val dynamicConfigService: DynamicConfigService
 ) {
   val log: Logger by lazy { LoggerFactory.getLogger(javaClass) }
 
-  fun userCanModifyResource(name: String): Boolean =
-    try {
-      val resource = repository.getResource(name)
-      userCanModifySpec(resource.serviceAccount, resource.id)
-    } catch (e: NoSuchResourceException) {
-      // If resource doesn't exist return true so a 404 is propagated from the controller.
-      true
-    }
+  enum class Action {
+    READ, WRITE;
+    override fun toString() = name.toLowerCase()
+  }
 
-  fun userCanModifySpec(serviceAccount: String, specOrName: Any): Boolean {
-    val auth = SecurityContextHolder.getContext().authentication
-    return userCanAccessServiceAccount(auth, serviceAccount, "Resource: $specOrName")
+  enum class TargetEntity {
+    APPLICATION, DELIVERY_CONFIG, RESOURCE;
+    override fun toString() = name.toLowerCase()
   }
 
-  fun userCanModifyApplication(application: String): Boolean = try {
-    val deliveryConfig = repository.getDeliveryConfigForApplication(application)
-    userCanModifyDeliveryConfig(deliveryConfig.serviceAccount, application, deliveryConfig.name)
-  } catch (e: NoSuchElementException) {
-    // If there are no delivery configs for that application return true so this is handled correctly in the controller.
-    true
+  private fun enabled() = dynamicConfigService.isEnabled("keel.authorization", true)
+
+  /**
+   * Returns true if the caller has the specified permission (action) to access the application associated with the
+   * specified target object.
+   */
+  fun hasApplicationPermission(action: String, target: String, identifier: String) =
+    passes { hasApplicationPermission(Action.valueOf(action), TargetEntity.valueOf(target), identifier) }
+
+  /**
+   * Returns true if  the caller has access to the specified service account.
+   */
+  fun hasServiceAccountAccess(target: String, identifier: String) =
+    passes { hasServiceAccountAccess(TargetEntity.valueOf(target), identifier) }
+
+  /**
+   * Returns true if the caller has the specified permission (action) to access the cloud account associated with the
+   * specified target object.
+   */
+  fun hasCloudAccountPermission(action: String, target: String, identifier: String) =
+    passes { hasCloudAccountPermission(Action.valueOf(action), TargetEntity.valueOf(target), identifier) }
+
+  /**
+   * Verifies that the caller has the specified permission (action) to access the application associated with the
+   * specified target object.
+   *
+   * @throws AccessDeniedException if caller does not have the required permission.
+   */
```

Comment:
For these functions, I'd expect a function starting with `has` to return a `Boolean`, not throw an exception if `false`. What do you think about changing these to just return false if the thing doesn't have access, since you're just swallowing the exception in `passes {}` anyways? Then you could remove the `passes()` function, and I think this would get simpler.

Reply:
The reason I wrote it this was was so that these functions could be used from the body of controllers and would take care of creating the exception with the right information, as opposed to letting each controller do that (including because the controller won't have all the info).

You can see a usage example here:
https://github.com/spinnaker/keel/blob/e313b5bbb5c0ee9d1d3836cd930000092dab7c76/keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/InteractiveNotificationCallbackController.kt#L44

But I agree the name is bad. Will replace `has` with `verify` (as the comment says).

Reply:
Oh I see. I think since that specific controller seems like a special case I would have just thrown the exception there, instead of throwing it everywhere but swallowing it most places. Do you anticipate this method being called from controllers a lot?

Reply:
I had no idea if it will be common or not. I hope it won't, but it didn't seem to hurt to encapsulate that in `AuthorizationSupport`. I think we would be tempted to just copy/paste a "if this then throw exception" from one controller to another if I didn't, and end up with inconsistent error messages. From the point of view of consumers of this class, most uses should be using `@PreAuthorize` annotations with the `hasXyz` form and they shouldn't care about the implementation underneath.